src/NUnitEngine/nunit.engine.core/DotNetHelper.cs

```diff
@@ -2,65 +2,113 @@
+                        : "/usr/shared/dotnet/");
+            X86InstallDirectory = 
+                Environment.GetEnvironmentVariable("DOTNET_ROOT_X86") ?? (
+                    IsWindows
+                        ? (string)Registry.LocalMachine.OpenSubKey(X86_SUBKEY1)?.GetValue("InstallLocation") ??
+                          (string)Registry.LocalMachine.OpenSubKey(X86_SUBKEY2)?.GetValue("InstallLocation") ?? @"C:\Program Files (x86)\dotnet"
+                        : "/usr/shared/dotnet/");
+#pragma warning restore CA1416
+            Runtimes = new List<RuntimeInfo>();
+            foreach (string line in DotnetCommand("--list-runtimes"))
+            {
+                string[] parts = line.Trim().Split([' '], 3);
+                Runtimes.Add(new RuntimeInfo(parts[0], parts[1], parts[2].Trim(['[', ']'])));
+            }
+        }
+
+        /// <summary>
+        /// Get the correct install directory, depending on whether we need X86 or X64 architecture.
+        /// </summary>
+        /// <param name="x86">Flag indicating whether the X86 architecture is needed</param>
+        /// <returns></returns>
+        public static string GetInstallDirectory(bool x86) => x86
+            ? X86InstallDirectory : X64InstallDirectory;
+
+        public static IEnumerable<RuntimeInfo> GetRuntimes(string name) => Runtimes.Where(r => r.Name == name);
 
-            if (_x86InstallDirectory == null)
+        private static IEnumerable<string> DotnetCommand(string arguments)
+        {
+            var process = new Process
             {
-#if NETFRAMEWORK
-                if (Path.DirectorySeparatorChar == '\\')
-#else
-                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-#endif
+                StartInfo = new ProcessStartInfo
                 {
-                    RegistryKey key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\dotnet\SetUp\InstalledVersions\x86\");
-                    _x86InstallDirectory = (string)key?.GetValue("InstallLocation");
+                    FileName = "dotnet",
+                    Arguments = arguments,
+                    UseShellExecute = false,
+                    RedirectStandardOutput = true,
+                    CreateNoWindow = true
                 }
-                else
-                    _x86InstallDirectory = "/usr/shared/dotnet/";
+            };
+
+            try
+            {
+                process.Start();
+            }
+            catch (Exception)
+            {
+                // Failed to start dotnet command. Assume no versions are installed and just return
+                yield break;
             }
 
-            return _x86InstallDirectory;
+            while (!process.StandardOutput.EndOfStream)
+                yield  return process.StandardOutput.ReadLine();
         }
+
+#if NETFRAMEWORK
+        private static bool IsWindows => Path.DirectorySeparatorChar == '\\';
```

Comment:
I like this condition as I don't think any other OS flipped the path separator, but in other places `Environment.OSVersion.Platform == PlatformID.Win32NT` is used.

Reply:
Good idea.